File: frontend/src/api/entries.ts
import type { ResourceResponse, SynthesizedDictionaryEntry } from '@/types/api';
import { api, transformError } from './core';

export interface EntryUpdateRequest {
  status?: string;
  tags?: string[];
  metadata?: Record<string, any>;
}

export interface ImageAddRequest {
  image_ids: string[];
}

export interface RegenerateRequest {
  components?: Set<string>;
  force?: boolean;
}

export interface FieldSelection {
  include?: string[];
  exclude?: string[];
  expand?: string[];
}

export const entriesApi = {
  // Get synthesized dictionary entry by ID - GET /words/entries/{entry_id}
  async getEntry(
    entryId: string,
    options?: FieldSelection
  ): Promise<SynthesizedDictionaryEntry> {
    try {
      const params: Record<string, any> = {};
      
      if (options?.include?.length) params.include = options.include.join(',');
      if (options?.exclude?.length) params.exclude = options.exclude.join(',');
      if (options?.expand?.length) params.expand = options.expand.join(',');
      
      const response = await api.get<ResourceResponse>(
        `/words/entries/${entryId}`,
        { params }
      );
      
      return response.data.data;
    } catch (error) {
      throw transformError(error);
    }
  },

  // Update synthesized dictionary entry - PUT /words/entries/{entry_id}
  async updateEntry(
    entryId: string, 
    updates: EntryUpdateRequest
  ): Promise<SynthesizedDictionaryEntry> {
    try {
      const response = await api.put<ResourceResponse>(
        `/words/entries/${entryId}`,
        updates
      );
      
      return response.data.data;
    } catch (error) {
      throw transformError(error);
    }
  },

  // Delete synthesized dictionary entry - DELETE /words/entries/{entry_id}
  async deleteEntry(entryId: string): Promise<boolean> {
    try {
      const response = await api.delete<ResourceResponse>(`/words/entries/${entryId}`);
      return response.data.data.deleted;
    } catch (error) {
      throw transformError(error);
    }
  },

  // Add images to synthesized dictionary entry - POST /words/entries/{entry_id}/images
  async addImagesToEntry(
    entryId: string,
    imageIds: string[]
  ): Promise<{
    entry_id: string;
    image_ids: string[];
    images_added: number;
    total_images: number;
  }> {
    try {
      const request: ImageAddRequest = { image_ids: imageIds };
      
      const response = await api.post<ResourceResponse>(
        `/words/entries/${entryId}/images`,
        request
      );
      
      return response.data.data;
    } catch (error) {
      throw transformError(error);
    }
  },

  // Remove image from synthesized dictionary entry - DELETE /words/entries/{entry_id}/images/{image_id}
  async removeImageFromEntry(
    entryId: string,
    imageId: string
  ): Promise<{
    entry_id: string;
    image_id: string;
    removed: boolean;
    remaining_images: number;
  }> {
    try {
      const response = await api.delete<ResourceResponse>(
        `/words/entries/${entryId}/images/${imageId}`
      );
      
      return response.data.data;
    } catch (error) {
      throw transformError(error);
    }
  },

  // Re-synthesize entry components - POST /words/entries/{entry_id}/regenerate
  async regenerateEntryComponents(
    entryId: string,
    options?: RegenerateRequest
  ): Promise<{
    entry_id: string;
    components: string[];
    regeneration_requested: boolean;
    status: string;
  }> {
    try {
      const request: RegenerateRequest = {
        components: options?.components || new Set(['all']),
        force: options?.force || false
      };

      const response = await api.post<ResourceResponse>(
        `/words/entries/${entryId}/regenerate`,
        request
      );
      
      return response.data.data;
    } catch (error) {
      throw transformError(error);
    }
  },

  // Utility methods for common operations
  
  // Refresh entry images (get updated entry with fresh image data)
  async refreshEntryImages(entryId: string): Promise<SynthesizedDictionaryEntry> {
    return entriesApi.getEntry(entryId, { expand: ['images'] });
  },

  // Update entry with image IDs and return updated entry
  async updateEntryWithImages(
    entryId: string, 
    imageIds: string[]
  ): Promise<SynthesizedDictionaryEntry> {
    // First add the images to the entry
    await entriesApi.addImagesToEntry(entryId, imageIds);
    
    // Then fetch the updated entry with expanded images
    return entriesApi.refreshEntryImages(entryId);
  },

  // Remove image and return updated entry
  async removeImageAndRefresh(
    entryId: string,
    imageId: string
  ): Promise<SynthesizedDictionaryEntry> {
    // Remove the image
    await entriesApi.removeImageFromEntry(entryId, imageId);
    
    // Return updated entry
    return entriesApi.refreshEntryImages(entryId);
  }
};

File: backend/src/floridify/api/routers/words/entries.py
"""
Synthesized Dictionary Entry API Router

Provides comprehensive CRUD operations for SynthesizedDictionaryEntry objects,
including component-level updates and image management.
"""

from typing import Any

from beanie import PydanticObjectId
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field

from ...core import (
    FieldSelection,
    NotFoundException,
    ResourceResponse,
    get_fields,
)
from ...repositories.synthesis_repository import SynthesisRepository, SynthesisUpdate

router = APIRouter()

# Request/Response Models
class EntryUpdateRequest(BaseModel):
    """Request model for updating synthesized entry metadata"""
    status: str | None = Field(None, description="Entry status")
    tags: list[str] | None = Field(None, description="Entry tags")
    metadata: dict[str, Any] | None = Field(None, description="Additional metadata")

class ComponentUpdateRequest(BaseModel):
    """Request model for updating entry components"""
    component_type: str = Field(..., description="Component type (definition, image, fact, etc.)")
    component_id: str = Field(..., description="Component ID to update")
    updates: dict[str, Any] = Field(..., description="Update data")

class RegenerateRequest(BaseModel):
    """Request model for re-synthesizing entry components"""
    components: set[str] = Field(default_factory=lambda: {"all"}, description="Components to regenerate")
    force: bool = Field(False, description="Force regeneration even if data exists")

class ImageAddRequest(BaseModel):
    """Request model for adding images to entry"""
    image_ids: list[PydanticObjectId] = Field(..., description="Image IDs to add to entry")

# Repository dependency
def get_synthesis_repo() -> SynthesisRepository:
    return SynthesisRepository()

# Routes

@router.get("/{entry_id}")
async def get_entry(
    entry_id: str,
    field_selection: FieldSelection = Depends(get_fields),
    repo: SynthesisRepository = Depends(get_synthesis_repo)
) -> ResourceResponse:
    """
    Get a synthesized dictionary entry by ID.
    
    Supports field selection for controlling which components are included:
    - expand: Load related components (definitions, images, facts, etc.)
    - include: Only include specified fields
    - exclude: Exclude specified fields
    """
    try:
        entry_dict = await repo.get_with_expansion(
            PydanticObjectId(entry_id),
            expand=field_selection.expand,
            include=field_selection.include,
            exclude=field_selection.exclude
        )
        
        if not entry_dict:
            raise NotFoundException(f"Entry with ID {entry_id} not found")
        
        # Update access tracking
        entry_oid = PydanticObjectId(entry_id)
        await repo.update_access_info(entry_oid)
        
        return ResourceResponse(
            data=entry_dict,
            metadata={
                "id": entry_id,
                "last_updated": entry_dict.get("updated_at"),
                "access_count": entry_dict.get("access_count", 0)
            }
        )
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid entry ID format: {entry_id}"
        )

@router.put("/{entry_id}")
async def update_entry(
    entry_id: str,
    request: EntryUpdateRequest,
    repo: SynthesisRepository = Depends(get_synthesis_repo)
) -> ResourceResponse:
    """Update synthesized entry metadata"""
    try:
        entry_oid = PydanticObjectId(entry_id)
        
        # Build update data
        update_data: dict[str, Any] = {}
        if request.status is not None:
            update_data["status"] = request.status
        if request.tags is not None:
            update_data["tags"] = request.tags
        if request.metadata is not None:
            update_data["metadata"] = request.metadata
        
        # Convert to SynthesisUpdate
        update_request = SynthesisUpdate(**update_data)
        updated_entry = await repo.update(entry_oid, update_request)
        
        return ResourceResponse(
            data=updated_entry.model_dump(mode="json"),
            metadata={
                "id": str(updated_entry.id),
                "updated_at": updated_entry.updated_at.isoformat() if updated_entry.updated_at else None
            }
        )
    except NotFoundException:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry with ID {entry_id} not found"
        )
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid entry ID format: {entry_id}"
        )

@router.delete("/{entry_id}")
async def delete_entry(
    entry_id: str,
    repo: SynthesisRepository = Depends(get_synthesis_repo)
) -> ResourceResponse:
    """Delete a synthesized dictionary entry"""
    try:
        entry_oid = PydanticObjectId(entry_id)
        await repo.delete(entry_oid)
        
        return ResourceResponse(
            data={"deleted": True},
            metadata={"id": entry_id}
        )
    except NotFoundException:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry with ID {entry_id} not found"
        )
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid entry ID format: {entry_id}"
        )

# Image Management Routes (Primary Focus)

@router.post("/{entry_id}/images")
async def add_images_to_entry(
    entry_id: str,
    request: ImageAddRequest,
    repo: SynthesisRepository = Depends(get_synthesis_repo)
) -> ResourceResponse:
    """
    Add images to a synthesized dictionary entry.
    
    This is the primary endpoint for associating uploaded images with entries.
    """
    try:
        entry_oid = PydanticObjectId(entry_id)
        
        # Image IDs are already PydanticObjectIds
        image_oids = request.image_ids
        
        # Get current entry
        entry = await repo.get(entry_oid)
        if not entry:
            raise NotFoundException(f"Entry with ID {entry_id} not found")
        
        # Add images to entry (avoid duplicates)
        current_image_ids = set(entry.image_ids or [])
        new_image_ids = current_image_ids.union(set(image_oids))
        
        # Update entry with new images
        update_request = SynthesisUpdate(image_ids=list(new_image_ids))
        updated_entry = await repo.update(entry_oid, update_request)
        
        return ResourceResponse(
            data={
                "entry_id": entry_id,
                "image_ids": [str(img_id) for img_id in updated_entry.image_ids],
                "images_added": len(image_oids),
                "total_images": len(updated_entry.image_ids)
            },
            metadata={
                "updated_at": updated_entry.updated_at.isoformat() if updated_entry.updated_at else None
            }
        )
    except NotFoundException:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry with ID {entry_id} not found"
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid ID format: {str(e)}"
        )

@router.delete("/{entry_id}/images/{image_id}")
async def remove_image_from_entry(
    entry_id: str,
    image_id: str,
    repo: SynthesisRepository = Depends(get_synthesis_repo)
) -> ResourceResponse:
    """Remove an image from a synthesized dictionary entry"""
    try:
        entry_oid = PydanticObjectId(entry_id)
        image_oid = PydanticObjectId(image_id)
        
        # Get current entry
        entry = await repo.get(entry_oid)
        if not entry:
            raise NotFoundException(f"Entry with ID {entry_id} not found")
        
        # Remove image from entry
        current_image_ids = entry.image_ids or []
        updated_image_ids = [img_id for img_id in current_image_ids if img_id != image_oid]
        
        # Update entry
        update_request = SynthesisUpdate(image_ids=updated_image_ids)
        updated_entry = await repo.update(entry_oid, update_request)
        
        return ResourceResponse(
            data={
                "entry_id": entry_id,
                "image_id": image_id,
                "removed": True,
                "remaining_images": len(updated_entry.image_ids)
            },
            metadata={
                "updated_at": updated_entry.updated_at.isoformat() if updated_entry.updated_at else None
            }
        )
    except NotFoundException:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry with ID {entry_id} not found"
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid ID format: {str(e)}"
        )

@router.post("/{entry_id}/regenerate")
async def regenerate_entry_components(
    entry_id: str,
    request: RegenerateRequest,
    repo: SynthesisRepository = Depends(get_synthesis_repo)
) -> ResourceResponse:
    """
    Re-synthesize components of a dictionary entry.
    
    This endpoint triggers AI regeneration of specific components
    or the entire entry if requested.
    """
    try:
        entry_oid = PydanticObjectId(entry_id)
        
        # Get current entry
        entry = await repo.get(entry_oid)
        if not entry:
            raise NotFoundException(f"Entry with ID {entry_id} not found")
        
        # TODO: Implement actual regeneration logic
        # This would involve calling the synthesis system to regenerate  
        # the requested components using the synthesizer's regenerate_entry_components method
        
        return ResourceResponse(
            data={
                "entry_id": entry_id,
                "components": list(request.components),
                "regeneration_requested": True,
                "status": "pending"
            },
            metadata={
                "requested_at": "2025-01-31T00:00:00Z"  # TODO: Use actual timestamp
            }
        )
    except NotFoundException:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Entry with ID {entry_id} not found"
        )
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid entry ID format: {entry_id}"
        )